Q24ECU/core/include/hal/hal_spi.h

```diff
@@ -0,0 +1,304 @@
+    uint8_t SPI_SSM;  // Slave Management (Software or Hardware)
+} SPI_Config_t;
+
+// Handle structure for SPIx peripheral
+typedef struct
+{
+    SPI_TypeDef *pSPIx; // Holds the base address of the SPI peripheral. Struct overlay for the SPI registers.
+    SPI_Config_t SPIConfig;
+    uint8_t *pTxBuffer;
+    uint8_t *pRxBuffer;
+    uint32_t TxLen;
+    uint32_t RxLen;
+    uint8_t TxState;
+    uint8_t RxState;
+} SPI_Handle_t;
+
+// Peripheral clocks setup
+static inline void HAL_SPI_PeriClockControl(SPI_TypeDef *pSPIx, bool enable)
+{
+    if (enable)
+    {
+        if (pSPIx == SPI1)
+        {
+            SPI1_PCLK_EN();
+        }
+        else if (pSPIx == SPI2)
+        {
+            SPI2_PCLK_EN();
+        }
+        else if (pSPIx == SPI3)
+        {
+            SPI3_PCLK_EN();
+        }
+        else if (pSPIx == SPI4)
+        {
+            SPI4_PCLK_EN();
+        }
+    }
+    else
+    {
+        if (pSPIx == SPI1)
+        {
+            SPI1_PCLK_DI();
+        }
+        else if (pSPIx == SPI2)
+        {
+            SPI2_PCLK_DI();
+        }
+        else if (pSPIx == SPI3)
+        {
+            SPI3_PCLK_DI();
+        }
+        else if (pSPIx == SPI4)
+        {
+            SPI4_PCLK_DI();
+        }
+    }
+}
+
+// Init and DeInit
+void HAL_SPI_Init(SPI_Handle_t *pSPIHandle)
+{
+    // First configure SPI_CR1 register
+    uint32_t tempReg = 0;
+
+    // 1. Configure device mode
+    tempReg |= pSPIHandle->SPIConfig.SPI_DeviceMode << SPI_CR1_MSTR_Pos;
+
+    // 2. Configure bus config
+    if (pSPIHandle->SPIConfig.SPI_BusConfig == SPI_BUS_CONFIG_FD)
+    {
+        // BIDI mode should be cleared
+        tempReg &= ~(SPI_CR1_BIDIMODE);
+    }
+    else if (pSPIHandle->SPIConfig.SPI_BusConfig == SPI_BUS_CONFIG_HD)
+    {
+        // BIDI mode should be set
+        tempReg |= SPI_CR1_BIDIMODE;
+    }
+    else if (pSPIHandle->SPIConfig.SPI_BusConfig == SPI_BUS_CONFIG_SIMPLEX_RXONLY)
+    {
+        // BIDI mode should be cleared
+        tempReg &= ~(SPI_CR1_BIDIMODE);
+        // RXONLY bit must be set
+        tempReg |= SPI_CR1_RXONLY;
+    }
+
+    // 3. Configure the SPI serial clock speed (baud rate)
+    tempReg |= pSPIHandle->SPIConfig.SPI_SclkSpeed << SPI_CR1_BR_Pos;
+
+    // 4. Configure the DFF
+    tempReg |= pSPIHandle->SPIConfig.SPI_DFF << SPI_CR1_DFF_Pos;
+
+    // 5. Configure the polarity
+    tempReg |= pSPIHandle->SPIConfig.SPI_CPOL << SPI_CR1_CPOL_Pos;
+
+    // 6. Configure the phase
+    tempReg |= pSPIHandle->SPIConfig.SPI_CPHA << SPI_CR1_CPHA_Pos;
+
+    pSPIHandle->pSPIx->CR1 = tempReg;
+}
+
+void HAL_SPI_DeInit(SPI_TypeDef *pSPIx)
+{
+    if (pSPIx == SPI1)
+    {
+        SPI1_REG_RESET();
+    }
+    else if (pSPIx == SPI2)
+    {
+        SPI2_REG_RESET();
+    }
+    else if (pSPIx == SPI3)
+    {
+        SPI3_REG_RESET();
+    }
+    else if (pSPIx == SPI4)
+    {
+        SPI4_REG_RESET();
+    }
+}
+
+// Data send and receive - blocking call
```

Comment:
maybe this applies maybe it doesn't but some sort of a `bool send_ready()` function might be nice

Reply:
added